### PR TITLE
FIFO in Calyx

### PR DESCRIPTION
**Pull request by anshumanmohan**
This is a first step towards the work we've sketched out in #1496. 

The plan may have drifted a little from the producer/consumer example laid out right up top in the issue. I'd love to chat about getting it back to where we wanted. Regardless, this is a quick proof of concept of FIFOs realized using circular queues.

It's rather janky Calyx, so thoughts about style are much appreciated. If there's a feature that you think should be used but I seem not to have used it, it's likely not an opinionated choice! It's probably because I don't know about the feature.

The way to run/test it also pretty janky: you only really get to look into the state of affairs via the memories `ans` and `err`, so you need to comment out later `invoke` statements if you want a look into the status at a particular point in time.

I've been running it with
```
fud e --from calyx --to jq --through verilog --through dat -s verilog.data fifo.data -s jq.expr ".memories" -q fifo.futil
```

I'm just surfacing this for your comments, and meanwhile I'll start writing eDSL code that generates this same handwritten Calyx!

Update: now that the handwritten version is gone, run this with:
```
python3 fifo.py > fifo.futil
fud e --from calyx --to jq --through verilog --through dat -s verilog.data fifo.data -s jq.expr ".memories" -q fifo.futil
```

### CODE REVIEW

**Comment by anshumanmohan** [comment]
Update: over the day, I have gone ahead and also written eDSL code that generates the same stuff as the handwritten code. It is not pretty, and will likely change a lot as the Calyx code itself changes, but I just wanted a working framework going.

File: calyx-py/test/correctness/fifo.futil [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    // simulate a circular queue of size 10
+    full = std_reg(1); // This will be 1 if the queue is full, 0 otherwise
+    empty = std_reg(1); // This will be 1 if the queue is empty, 0 otherwise
+
+    ref ans = std_mem_d1(32, 1, 32);
+    // If the user wants to pop, we will write the popped value to `ans`
+    ref err = std_mem_d1(1, 1, 1);
+    // We'll raise this in case of overflow, underflow, or if the user calls pop and push at the same time
+
+    eq0 = std_eq(1); // We'll use this to check if pop == push
```
Comment (rachitnigam) [inline calyx-py/test/correctness/fifo.futil:18]: Since these `eq` have a defined role, I'd probably give them more meaningful names like `pop_eq_push` or something?

**Comment by rachitnigam** [comment]
This is looking really cool to me! @paili0628 would something like this be useful for the HiSparse implementation?

**Comment by rachitnigam** [comment]
And yeah, the eDSL seems like a good idea, especially if it lets us stress test the FIFO implementation and make sure it works for all corner cases?

**Comment by anshumanmohan** [comment]
Thanks for taking a look! I'm away from my work computer now, but for the next steps I propose:
- Drop the handwritten version.
- Tweak the eDSL version to take your suggestion re: `eq`-cell names.
- Rework the I/O interface a little so that it is not quite so janky. For instance, the .data file could contain an input memory such as `[100, 101, -1, 102]`, which is read as "push 100, push 101, pop, push 102". The output memory cells would be used to show the pops, in order, until the first time we hit an error. In the "normal" case the output will be the same as the input, but see below.

After this I'd like to move on to PIFOs. There the output will begin to become interesting: a _permutation_, possibly the identity permutation, of the input.

File: calyx-py/test/correctness/fifo.py [hunk 2]
```diff
@@ -247,57 +290,93 @@ def add_main(prog, fifo):
...
+                        fifo,
+                        in_pop=cb.const(1, 0),
+                        in_push=cb.const(1, 1),
+                        in_payload=command.out,
+                        ref_ans=ans,
+                        ref_err=err,
+                    ),
+                ),
+                incr_i,  # Increment the command index
+                # cb.if_(
```
Comment (anshumanmohan) [inline calyx-py/test/correctness/fifo.py:370]: These lines are commented out for a similar reason. Same issue: I want to access a memory that is already of interest to the while loop. In this case I want to write to it.

File: calyx-py/test/correctness/fifo.py [hunk 3]
```diff
@@ -247,57 +290,93 @@ def add_main(prog, fifo):
...
+                        ),
+                        cb.if_(  # but was it successful?
+                            err_eq_zero[0].out,
+                            err_eq_zero[1],
+                            [  # yes, it was successful: write the answer to ans_mem
+                                write_ans,
+                                incr_j,
+                            ],
+                        ),
+                        # AM:
```
[the remark this answers is not in the document]
Comment (anshumanmohan) [reply calyx-py/test/correctness/fifo.py:355]: It sounds like it _can_ be a register (see #1616) which would fix this issue I think. Regardless, I sorta don't get the issue. These are reads from the same memory, yes, but are not _parallel_ reads as far as I can tell. Either I'm missing something and Calyx is saving me, or Calyx is being overly cautious here...

[the remark this answers is not in the document]
Comment (rachitnigam) [reply calyx-py/test/correctness/fifo.py:355]: I suspect that you're running afoul of [the stability check](https://docs.calyxir.org/lang/ref.html#while) which assumes that the `comb group` is active throughout the body of the `if` or the `while`. It is a little annoying but we ended up figuring out that this restricts the language enough to reject some malformed programs (see https://github.com/cucapra/calyx/discussions/927).

File: calyx-py/test/correctness/fifo.py (hunk 2 above)
Comment (rachitnigam) [reply calyx-py/test/correctness/fifo.py:370]: Are you using `std_mem` or `seq_mem`. If you're using `std_mem`, I recommend switching to `seq_mem` since we'll deprecate the former soon (#1261) because they are not easily mapped onto good physical designs.

File: calyx-py/test/correctness/fifo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (anshumanmohan) [reply calyx-py/test/correctness/fifo.py:355]: Gotcha, in this case I do need this behavior, so I'll just have to convert that `comb group` into a regular `group`. That is, instead of 
```
if err_eq_0.out with err_eq_0_group {...}
``` 

I'll change `err_eq_0_group` so it is not `comb` and then have 
```
[ 
  err_eq_0_group; if err_eq_0.out {...}
]
```
 in all instances. Does that sound kosher?

And actually now that I know the reason for the error, I'm guessing that moving to registers would not have helped, huh?

[the remark this answers is not in the document]
Comment (rachitnigam) [reply calyx-py/test/correctness/fifo.py:355]: Yeah, I think that should be fine! This is a general pattern that we use for conditions and is in fact exactly how `with` statements for `if` and `while` are compiled.

[the remark this answers is not in the document]
Comment (sampsyo) [reply calyx-py/test/correctness/fifo.py:355]: Yep, confirming: the usual way we do things is that the `with` group is truly combinational; it should just do combinational operations like `std_lt` or whatever on stable values. Reading from a memory (by assigning to its address port) is not combinational. So the pattern is: read from a memory in a proper Calyx group activation, put that in a register, and then use a combinational group if necessary to compute stuff about that value you read.

[the remark this answers is not in the document]
Comment (rachitnigam) [reply calyx-py/test/correctness/fifo.py:355]: > Reading from a memory (by assigning to its address port) is not combinational.

This is true for `seq_mem`, which have a sequential interface for both reads and writes but not true for `std_mem` which have a combinational read interface and sequential write interface

**Comment by anshumanmohan** [comment]
I've done some cleanup since you both last saw this. There remain some issues that I have thrashed about on a bit, and for those I'm seeking slightly more hands-on help.

### Using `seq_mem` correctly. 

I've tried to replace the `mem_d1` cells with `seq_mem_d1` cells, and figured out that some other groups' `done` signals need to rely on `read_done` and `write_done`, not just `done`. However, this led to:
    1. strange blocking behavior when I tried to make the groups rely on multiple done conditions, using the `(cell1.done & cell2.read_done ) @ 1` syntax.
    2. no writes being registered to the `seq_mem` when I dropped this careful `done` signal. That is, my answer memory was just coming out as it went in: all zeroes.

### Getting my main loop to run. 

Let me write out the goal straightforwardly.

The FIFO component listens for the commands `push` and `pop`.
- If `pop` is called on it successfully, it returns the popped value in `ans`. 
- If `push` is called, it expects a value in `payload`, and it tries to push it in.
- The fifo method can raise an `err` flag in case of error.

I just want my main loop to run from 0 to 15, pulling commands off a 15-long `command` memory and passing them to the FIFO component. If the command is `pop` and it succeeds, it writes the popped value to an `answer` memory. If the FIFO component ever has an error, I want the loop to break. The `command` and `answer` memories are external, and they are how we communicate with the user.

Right now I have the loop running contingent on the `err` flag being low. This `err` flag is passed by reference to the FIFO component, and the FIFO component may raise it during its work. So far so good! Indeed, if an error is raised before we run out of commands, everything seems fine.

However, what about the "normal" case where there is no error? I wanted to set things up so that, if the command index ever became 15, we would raise the very same error flag. This is proving very tricky. Note that, since you reviewed the PR, the flag has gone from being a 1-cell memory to a register.

Another small thing is that, when `pop` fails, we should not be writing its answer into the memory. This issue is what we discussed [above](https://github.com/cucapra/calyx/pull/1613#discussion_r1267300225), but in the code that is currently checked in, this issue is somewhat suppressed: I just got the FIFO component to zero out the `ans` register in case of an error.

Both of these issues are marked with "AM" in the code I've checked in.

**Comment by anshumanmohan** [comment]
Quick update: I noticed that passing cells around by reference was a workaround to the issue of combinational groups blocking off cells, so I leaned into it. It's a super duper hack, and I would love to massage it away. 

I just have a further group whose whole job is to raise the `err` flag if `i=15`. It takes `err` by reference as `i` as just a plain input. This is run as the last step in the loop body every time. This works for now.

**Comment by rachitnigam** [comment]
@anshumanmohan not sure if the `seq_mem` issue was resolved or not but here is the general guideline to using a `seq_mem`:
- The `read_en` interface implements a *sequential* reads which takes a cycle to return the value. In order to get the right value, you need to depend on `read_done`. I think you're already doing this.
- The is *latched* after a `read_done` operation which means it is available *outside* the group that did the read as long as another `read_en` or `write_en` is not triggered.

So, if you want to perform a read from the memory, you can write the following:
```
group do_read {
  mem.addr0 = ...
  mem.read_en = ...
  do_read[done] = mem.read_done
}
group write_reg {
  reg.in = mem.out;
  reg.write_en = 1'd1
  write_reg[done] = reg.done
}

control {
  seq { do_read; write_reg; }
}
```

**Comment by anshumanmohan** [comment]
Many thanks! I didn't fully realize that I should just make two groups and orchestrate them in sequence. It works now, and I've moved over to `seq_mem`. I've also added asserts to enforce the type of the memory.

This PR now suffers no fatal flaws, except that its main loop is a little janky. Both of these are marked with "AM" in the code.
1. If a `pop` fails, we should not write the popped value to the answer list.
2. After we hit `i=15`, we should break the loop more elegantly than we do now.

These issues are both suppressed right now in hacky ways:
1. When the fifo component returns an error, it zeroes out its answer cell. So, although the main loop _does_ write an invalid answer to its answer list, that invalid answer is `0` and therefore does not cause confusion.
2. I just have a bespoke component `raise_err_if_i_eq_15` whose whole job is to potentially raise the `err` flag. This works fine, it's just overkill and there probably exists a nicer way.
